metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java

```diff
@@ -240,7 +240,7 @@ boolean check() {
     /**
      * The broker heartbeat manager, or null if this controller is on standby.
      */
-    private BrokerHeartbeatManager heartbeatManager;
+    private volatile BrokerHeartbeatManager heartbeatManager;
```

Comment:
Hm, I guess this was previously a bug technically. Glancing through the usages, we may have some TOCTOU bugs with this. We tend to check if it's null and then do some stuff, but maybe it could become null after that check if there was a race on the variable read.

Reply:
As far as I know, it wasn't a TOCTOU previously since all the reads and writes were done from the main controller thread. But now we added a pathway directly from the request handler thread pool, so it needs to be volatile.

We should try to run helgrind on Kafka some day...